modules/phase_field/include/materials/ParsedMaterialHelper.h

```diff
@@ -250,4 +263,173 @@ ParsedMaterialHelper<T>::computeProperties()
+  const std::string & getSymbolName() const { return _fparser_name; };
+
+  /// get the property name
+  const std::string getPropertyName() const
+  {
+    mooseAssert( _parent_material != NULL, "_parent_material pointer is NULL" );
+    return _parent_material->propertyName(_base_name, _derivative_vars);
+  };
+
+  /// get the property reference
+  const MaterialProperty<Real> & value() const
+  {
+    mooseAssert( _value != NULL, "_value pointer is NULL" );
+    return *_value;
+  }
+
+private:
+  void parseDerivative(const std::string &);
+  void parseDependentVariables(const std::string &);
+
+  /// name used in function expression
+  std::string _fparser_name;
+
+  /// function material property base name
+  std::string _base_name;
+
+  std::vector<std::string> _dependent_vars;
+  std::vector<std::string> _derivative_vars;
+
+  /// material property value
+  MaterialProperty<Real> * _value;
+
+  /// parent material class
+  T * _parent_material;
+};
+
+template<class T>
+ParsedMaterialHelper<T>::FunctionMaterialPropertyDescriptor::FunctionMaterialPropertyDescriptor(const std::string & expression, T * parent) :
+    _dependent_vars(),
+    _derivative_vars(),
+    _parent_material(parent)
+{
+  size_t define = expression.find_last_of(":=");
+
+  // expression contains a ':='
+  if (define != std::string::npos)
+  {
+    // section before ':=' is the name used in the function expression
+    _fparser_name = expression.substr(0, define);
+
+    // parse right hand side
+    parseDerivative(expression.substr(define+2));
+  }
+  else
+  {
+    // parse entire expression and use natural material property base name
+    // for D(x(t),t,t) this would simply be 'x'!
+    parseDerivative(expression);
+    _fparser_name = _base_name;
+  }
+
+  // get the material property reference
+  _value = &(parent->template getMaterialProperty<Real>(getPropertyName()));
+}
+
+template<class T>
+ParsedMaterialHelper<T>::FunctionMaterialPropertyDescriptor::FunctionMaterialPropertyDescriptor() :
+    _value(NULL),
+    _parent_material(NULL)
+{
+}
+
+template<class T>
+ParsedMaterialHelper<T>::FunctionMaterialPropertyDescriptor::FunctionMaterialPropertyDescriptor(const FunctionMaterialPropertyDescriptor & rhs) :
+    _fparser_name(rhs._fparser_name),
+    _base_name(rhs._base_name),
+    _dependent_vars(rhs._dependent_vars),
+    _derivative_vars(rhs._derivative_vars),
+    _value(rhs._value),
+    _parent_material(rhs._parent_material)
+{
+}
+
+template<class T>
+void
+ParsedMaterialHelper<T>::FunctionMaterialPropertyDescriptor::parseDerivative(const std::string & expression)
+{
+  size_t open  = expression.find_first_of("[");
+  size_t close = expression.find_last_of("]");
+
+  if (open == std::string::npos && close == std::string::npos)
+  {
+    // no derivative requested
+    parseDependentVariables(expression);
+    return;
+  }
+  else if (open != std::string::npos && close != std::string::npos && expression.substr(0, open) == "D")
```

Comment:
Uhm yes. For example in malformed strings. And the expression empty test is
not testing the content between the brackets only.

On Fri, Apr 24, 2015, 12:42 John W. Peterson <notifications@github.com>
wrote:

> In modules/phase_field/include/materials/ParsedMaterialHelper.h
> <https://github.com/idaholab/moose/pull/4982#discussion_r29074510>:
>
> > +}
> > +
> > +template<class T>
> > +void
> > +ParsedMaterialHelper<T>::FunctionMaterialPropertyDescriptor::parseDerivative(const std::string & expression)
> > +{
> > +  size_t open  = expression.find_first_of("[");
> > +  size_t close = expression.find_last_of("]");
> > +
> > +  if (open == std::string::npos && close == std::string::npos)
> > +  {
> > +    // no derivative requested
> > +    parseDependentVariables(expression);
> > +    return;
> > +  }
> > +  else if (open != std::string::npos && close != std::string::npos && expression.substr(0, open) == "D")
>
> Is open ever != 1?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/4982/files#r29074510>.
>